src/main/java/pl/sda/refactoring/service/OrderService.java

```diff
@@ -0,0 +1,156 @@
+                          @NonNull List<OrderItem> orderItems,
+                          @NonNull Currency currency,
+                          String coupon) {
+        boolean exists = customerService.exists(customerId);
+        if (exists) {
+            if (!orderItems.isEmpty()) {
+                // exchange currencies in order items
+                List<OrderItem> exchItems = new ArrayList<>();
+                for (var item : orderItems) {
+                    var exchItem = item.toBuilder()
+                        .id(UUID.randomUUID())
+                        .exchPrice(currencyService.exchange(item.price(), item.currency(), currency))
+                        .build();
+                    exchItems.add(exchItem);
+                }
+
+                // calculate delivery
+                var tp = BigDecimal.ZERO;
+                var twInGrams = 0;
+                for (var item : exchItems) {
+                    tp = tp.add(item
+                            .exchPrice()
+                            .multiply(BigDecimal.valueOf(item.quantity())))
+                        .setScale(2, HALF_UP);
+                    switch (item.weightUnit()) {
+                        case GM -> twInGrams += item.weight();
+                        case KG -> twInGrams += item.weight() * 1000;
+                        case LB -> twInGrams += item.weight() * 453.59237;
+                    }
+                }
+                var deliveryPrice = BigDecimal.ZERO;
+                // delivery costs are in USD
+                var tpInUsd = currencyService.exchange(tp, currency, Currency.USD);
+                if (tpInUsd.compareTo(new BigDecimal("400.00")) > 0 && twInGrams < 2000) {
+                    deliveryPrice = new BigDecimal("0.00"); // free delivery
+                } else if (tpInUsd.compareTo(new BigDecimal("200.00")) > 0 && twInGrams < 1000) {
+                    deliveryPrice = new BigDecimal("12.00"); // first level
+                } else if (tpInUsd.compareTo(new BigDecimal("100.00")) > 0 && twInGrams < 1000) {
+                    deliveryPrice = new BigDecimal("15.00"); // second level
+                } else {
+                    // default delivery
+                    deliveryPrice = new BigDecimal("20.00");
+                }
+                deliveryPrice = currencyService.exchange(deliveryPrice, Currency.USD, currency);
+
+                // calculate discount
+                var discountPrice = BigDecimal.ZERO;
+                if (coupon != null) {
+                    var discount = discountService.getDiscount(coupon);
+                    if (discount != null) {
+                        discountPrice = tp
+                            .multiply(BigDecimal.valueOf(discount.value()))
+                            .setScale(2, HALF_UP);
+                        discountService.deactivate(coupon, customerId);
+                    }
+                }
+                if (freeDeliveryDays.contains(LocalDate.now())) {
```

Comment:
Create a service checking whether it is free delivery date instead of creating this object with list of dates passed in a constructor.